src/StreamJsonRpc.Tests/JsonRpcTests.cs

```diff
@@ -803,6 +812,114 @@ public async Task InvokeWithParameterObject_DefaultParameters()
         Assert.Equal(12, sum);
     }
 
+    [Fact]
+    public async Task InvokeWithParameterObject_ProgressParameter()
+    {
+        int report = 0;
+        ProgressWithCompletion<int> progress = new ProgressWithCompletion<int>(n =>
+        {
+            report = n;
+        });
+
+        int result = await this.clientRpc.InvokeWithParameterObjectAsync<int>(nameof(Server.MethodWithProgressParameter), new { p = progress }, this.TimeoutToken);
+
+        Assert.Equal(1, report);
+        Assert.Equal(1, result);
+    }
+
+    [Fact]
+    public async Task InvokeWithParameterObject_ProgressParameterMultipleRequests()
+    {
+        int report1 = 0;
+        ProgressWithCompletion<int> progress1 = new ProgressWithCompletion<int>(n =>
+        {
+            report1 = n;
+        });
+
+        int report2 = 0;
+        ProgressWithCompletion<int> progress2 = new ProgressWithCompletion<int>(n =>
+        {
+            report2 = n * 2;
+        });
+
+        int report3 = 0;
+        ProgressWithCompletion<int> progress3 = new ProgressWithCompletion<int>(n =>
+        {
+            report3 = n * 3;
+        });
+
+        await this.InvokeMethodWithProgressParameter(progress1);
+        await this.InvokeMethodWithProgressParameter(progress2);
+        await this.InvokeMethodWithProgressParameter(progress3);
+
+        await progress1.WaitAsync();
+        await progress2.WaitAsync();
+        await progress3.WaitAsync();
+
+        Assert.Equal(1, report1);
+        Assert.Equal(2, report2);
+        Assert.Equal(3, report3);
+    }
+
+    [Fact]
+    public async Task InvokeWithParameterObject_InvalidParamMethod()
```

Comment:
>InvokeWithParameterObject_InvalidParamMethod [](start = 22, length = 44)

I think the exception thrown is reasonable. Can you check the trace output of the test and share what it prints? I'm hoping on the server-side it gives a clear idea of why the method expected to take this wasn't selected.

Reply:
Sure, this is the output of the test:
Server Information: 3 : Listening started.
Client Information: 3 : Listening started.
Client Information: 8 : {"id":2,"method":"MethodWithInvalidProgressParameter"}
Client Verbose: 8 : Sent: {
  "jsonrpc": "2.0",
  "method": "MethodWithInvalidProgressParameter",
  "params": {
    "p": {}
  },
  "id": 2
}
Server Information: 7 : {"id":2,"method":"MethodWithInvalidProgressParameter"}
Server Verbose: 7 : Received: {
  "jsonrpc": "2.0",
  "method": "MethodWithInvalidProgressParameter",
  "params": {
    "p": 0
  },
  "id": 2
}
Server Information: 6 : Received request "2" for method "MethodWithInvalidProgressParameter".
Server Warning: 5 : Invocation of "MethodWithInvalidProgressParameter" cannot occur because arguments do not match any registered target methods.
Server Information: 8 : {"id":2,"error":{"code":-32602,"message":"Unable to find method 'MethodWithInvalidProgressParameter/1' on {no object} for the following reasons: An argument was not supplied for a required parameter."}}
Server Verbose: 8 : Sent: {
  "jsonrpc": "2.0",
  "error": {
    "code": -32602,
    "message": "Unable to find method 'MethodWithInvalidProgressParameter/1' on {no object} for the following reasons: An argument was not supplied for a required parameter.",
    "data": null
  },
  "id": 2
}
Client Information: 7 : {"id":2,"error":{"code":-32602,"message":"Unable to find method 'MethodWithInvalidProgressParameter/1' on {no object} for the following reasons: An argument was not supplied for a required parameter."}}
Client Verbose: 7 : Received: {
  "jsonrpc": "2.0",
  "error": {
    "code": -32602,
    "message": "Unable to find method 'MethodWithInvalidProgressParameter/1' on {no object} for the following reasons: An argument was not supplied for a required parameter.",
    "data": null
  },
  "id": 2
}
Client Warning: 12 : Received error response for request 2: InvalidParams "Unable to find method 'MethodWithInvalidProgressParameter/1' on {no object} for the following reasons: An argument was not supplied for a required parameter.": 
Server Information: 13 : Connection closing (LocallyDisposed: Stream has been disposed). 
Client Information: 13 : Connection closing (LocallyDisposed: Stream has been disposed).

---
In reply to: [304388083](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r304388083) [](ancestors = 304388083)

Reply:
hmmm... that error is not super-great, given the client *did* supply a value. We might want to find the code that's emitting that message and see if we can fine-tune it to distinguish between "no value" and "incompatible value".

Reply:
I found this in the JsonRpc class:
`return new JsonRpcError
                    {
                        Id = request.Id,
                        Error = new JsonRpcError.ErrorDetail
                        {
                            Code = JsonRpcErrorCode.InvalidParams,
                            Message = targetMethod.LookupErrorMessage,
                        },
                    };`

The targetMethod.LookupErrorMessage contains the message that is being thrown in the Exception, maybe we could customize that message just like we do for the MethodNotfoundError:

`return new JsonRpcError
                    {
                        Id = request.Id,
                        Error = new JsonRpcError.ErrorDetail
                        {
                            Code = JsonRpcErrorCode.MethodNotFound,
                            Message = string.Format(CultureInfo.CurrentCulture, Resources.RpcMethodNameNotFound, request.Method),
                        },
                    };`

Reply:
Sure. If you make any change in this area please review the messages in the relevant cases to make sure it all makes sense. IMO this is optional, but a nice to have.

---
In reply to: [304577320](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r304577320) [](ancestors = 304577320)